File: hdar/cpp/src/unsafe_host.cpp
#include "hdar/unsafe_host.hpp"
#include <cstdlib>
#include <chrono>
#include <filesystem>

namespace fs = std::filesystem;

namespace hdar {

UnsafeHostProvider::UnsafeHostProvider(const std::string& root)
    : sandbox_root_(root) {
    fs::create_directories(sandbox_root_);
}

RuntimeRecord UnsafeHostProvider::materialize(
    const std::string& runtime_id,
    const std::string& workspace_path,
    const std::string& image,
    const std::string& cpu_limit,
    const std::string& memory_limit,
    const std::string& network_policy) {

    std::string dir = runtime_dir(runtime_id);
    fs::create_directories(dir);

    // Copy workspace into sandbox
    if (fs::exists(workspace_path) && fs::is_directory(workspace_path)) {
        for (const auto& entry : fs::recursive_directory_iterator(workspace_path)) {
            if (entry.is_regular_file()) {
                std::string rel = fs::relative(entry.path(), workspace_path).string();
                std::string dest = dir + "/" + rel;
                fs::create_directories(fs::path(dest).parent_path());
                fs::copy_file(entry.path(), dest, fs::copy_options::overwrite_existing);
            }
        }
    }

    RuntimeRecord rec;
    rec.provider = "unsafe-host";
    rec.runtime_id = runtime_id;
    rec.cpu_limit = cpu_limit;
    rec.memory_limit = memory_limit;
    rec.workspace_mount = dir;
    rec.network_policy = network_policy;
    rec.start_timestamp = epoch_seconds();
    rec.exists = true;

    runtimes_[runtime_id] = rec;
    return rec;
}

ExecutionResult UnsafeHostProvider::execute(
    const std::string& runtime_id,
    const std::string& operation_type,
    const std::string& command,
    int timeout) {

    ExecutionResult result;
    result.operation_type = operation_type;
    result.command = command;

    auto it = runtimes_.find(runtime_id);
    if (it == runtimes_.end() || !it->second.exists) {
        result.stderr_text = "runtime not found or not running";
        return result;
    }

    std::string dir = runtime_dir(runtime_id);
    std::string full_cmd = "cd " + dir + " && " + command + " 2>&1";

    auto start = std::chrono::steady_clock::now();

    FILE* pipe = popen(full_cmd.c_str(), "r");
    if (!pipe) {
        result.stderr_text = "popen failed";
        return result;
    }

    char buffer[4096];
    std::string output;
    while (fgets(buffer, sizeof(buffer), pipe) != nullptr)
        output += buffer;

    int status = pclose(pipe);
    result.exit_code = WEXITSTATUS(status);
    result.stdout_text = output;
    result.success = (result.exit_code == 0);

    auto end = std::chrono::steady_clock::now();
    result.duration_ms = std::chrono::duration<double, std::milli>(end - start).count();

    return result;
}

RuntimeRecord UnsafeHostProvider::stop(const std::string& runtime_id) {
    auto it = runtimes_.find(runtime_id);
    if (it == runtimes_.end())
        throw std::runtime_error("runtime not found: " + runtime_id);

    it->second.stop_timestamp = epoch_seconds();
    return it->second;
}

RuntimeRecord UnsafeHostProvider::destroy(const std::string& runtime_id) {
    auto it = runtimes_.find(runtime_id);
    if (it == runtimes_.end())
        throw std::runtime_error("runtime not found: " + runtime_id);

    std::string dir = runtime_dir(runtime_id);
    if (fs::exists(dir))
        fs::remove_all(dir);

    it->second.delete_timestamp = epoch_seconds();
    it->second.exists = false;

    JsonValue inspection = JsonValue::object();
    inspection["exists"] = JsonValue::boolean(false);
    inspection["provider"] = JsonValue::string("unsafe-host");
    inspection["runtime_id"] = JsonValue::string(runtime_id);
    inspection["sandbox_dir"] = JsonValue::string(dir);
    inspection["dir_exists"] = JsonValue::boolean(false);
    it->second.post_delete_inspection = inspection;

    return it->second;
}

JsonValue UnsafeHostProvider::inspect(const std::string& runtime_id) {
    auto it = runtimes_.find(runtime_id);
    JsonValue v = JsonValue::object();

    if (it == runtimes_.end()) {
        v["exists"] = JsonValue::boolean(false);
        v["provider"] = JsonValue::string("unsafe-host");
        v["runtime_id"] = JsonValue::string(runtime_id);
        return v;
    }

    bool dir_exists = fs::exists(runtime_dir(runtime_id));
    v["exists"] = JsonValue::boolean(it->second.exists && dir_exists);
    v["provider"] = JsonValue::string("unsafe-host");
    v["runtime_id"] = JsonValue::string(runtime_id);
    v["sandbox_dir"] = JsonValue::string(runtime_dir(runtime_id));
    v["dir_exists"] = JsonValue::boolean(dir_exists);
    return v;
}

std::vector<std::string> UnsafeHostProvider::list_runtimes() {
    std::vector<std::string> result;
    for (const auto& [id, rec] : runtimes_)
        if (rec.exists && fs::exists(runtime_dir(id)))
            result.push_back(id);
    return result;
}

} // namespace hdar

File: hdar/hdar_core/providers/unsafe_host.py
"""Unsafe host provider — development and testing only.

Executes operations directly on the host filesystem inside a
sandboxed directory. NOT a real isolated runtime. Used for
testing the lifecycle controller without a container runtime.

The provider receipt honestly records provider="unsafe-host" so
no one mistakes this for isolated execution.
"""

from __future__ import annotations

import os
import shutil
import subprocess
import time
from pathlib import Path
from typing import Optional

from .base import ProviderBase, RuntimeRecord, ExecutionResult


class UnsafeHostProvider(ProviderBase):
    """Development provider — no isolation. For testing only."""

    def __init__(self, sandbox_root: str):
        self.sandbox_root = Path(sandbox_root)
        self.sandbox_root.mkdir(parents=True, exist_ok=True)
        self._runtimes: dict[str, RuntimeRecord] = {}

    @property
    def name(self) -> str:
        return "unsafe-host"

    def materialize(
        self,
        runtime_id: str,
        workspace_path: str,
        image: str = "",
        cpu_limit: str = "2",
        memory_limit: str = "2g",
        network_policy: str = "none",
    ) -> RuntimeRecord:
        runtime_dir = self.sandbox_root / runtime_id
        runtime_dir.mkdir(parents=True, exist_ok=True)

        # Copy workspace into runtime
        ws = Path(workspace_path)
        if ws.exists():
            dest = runtime_dir / "workspace"
            if dest.exists():
                shutil.rmtree(dest)
            shutil.copytree(ws, dest)

        record = RuntimeRecord(
            provider=self.name,
            runtime_id=runtime_id,
            image_digest=image or "unsafe-host:latest",
            vm_identity=f"process-{os.getpid()}",
            cpu_limit=cpu_limit,
            memory_limit=memory_limit,
            workspace_mount=str(runtime_dir / "workspace"),
            network_policy=network_policy,
            start_timestamp=time.time(),
            exists=True,
        )
        self._runtimes[runtime_id] = record
        return record

    def execute(
        self,
        runtime_id: str,
        operation_type: str,
        command: str,
        timeout: int = 60,
    ) -> ExecutionResult:
        if runtime_id not in self._runtimes:
            return ExecutionResult(
                operation_type=operation_type,
                command=command,
                exit_code=-1,
                stderr="runtime not found",
                success=False,
            )

        ws = self.sandbox_root / runtime_id / "workspace"
        start = time.time()
        try:
            proc = subprocess.run(
                command,
                shell=True,
                cwd=str(ws),
                capture_output=True,
                text=True,
                timeout=timeout,
            )
            duration = (time.time() - start) * 1000
            return ExecutionResult(
                operation_type=operation_type,
                command=command,
                exit_code=proc.returncode,
                stdout=proc.stdout,
                stderr=proc.stderr,
                duration_ms=duration,
                success=proc.returncode == 0,
            )
        except subprocess.TimeoutExpired:
            return ExecutionResult(
                operation_type=operation_type,
                command=command,
                exit_code=-1,
                stderr=f"timeout after {timeout}s",
                duration_ms=timeout * 1000,
                success=False,
            )

    def stop(self, runtime_id: str) -> RuntimeRecord:
        if runtime_id not in self._runtimes:
            return RuntimeRecord(
                provider=self.name, runtime_id=runtime_id, exists=False
            )
        record = self._runtimes[runtime_id]
        record.stop_timestamp = time.time()
        return record

    def destroy(self, runtime_id: str) -> RuntimeRecord:
        if runtime_id not in self._runtimes:
            return RuntimeRecord(
                provider=self.name, runtime_id=runtime_id, exists=False
            )
        record = self._runtimes[runtime_id]
        runtime_dir = self.sandbox_root / runtime_id
        if runtime_dir.exists():
            shutil.rmtree(runtime_dir)
        record.delete_timestamp = time.time()
        record.exists = False
        record.post_delete_inspection = self.inspect(runtime_id)
        del self._runtimes[runtime_id]
        return record

    def inspect(self, runtime_id: str) -> dict:
        if runtime_id in self._runtimes:
            return {"exists": True, **self._runtimes[runtime_id].to_dict()}
        runtime_dir = self.sandbox_root / runtime_id
        return {
            "exists": runtime_dir.exists(),
            "provider": self.name,
            "runtime_id": runtime_id,
        }

    def list_runtimes(self) -> list:
        return list(self._runtimes.keys())
